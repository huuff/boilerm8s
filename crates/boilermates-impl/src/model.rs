use derive_more::{Constructor, IntoIterator};
use heck::ToPascalCase as _;
use indexmap::IndexMap;

nestify::nest! {
  // XXX: Use an indexmap so it's insertion order and deterministic
  #[derive(Default, IntoIterator)]
  pub struct OutputStructs(
    #[into_iterator(owned, ref, ref_mut)]
    #>[derive(Default)]
    pub IndexMap<String, pub struct OutputStruct {
      pub attributes: Vec<syn::Attribute>,
      #>[derive(Clone, Constructor)]
      pub fields: Vec<pub struct OutputField {
        pub definition: syn::Field,
        pub default: bool,
      }>,
  }>)
}

impl OutputStructs {
    /// Returns the names of all the output structs
    pub fn names(&self) -> impl Iterator<Item = &str> {
        self.0.keys().map(|it| it.as_str())
    }

    /// Initializes the output with the main struct and all the mates
    pub fn initialize(
        item: &syn::DeriveInput,
        attrs: proc_macro2::TokenStream,
    ) -> Result<Self, anyhow::Error> {
        use syn::parse::Parser as _;

        let mut output = Self::default();

        output.0.insert(
            item.ident.to_string(),
            OutputStruct {
                fields: Default::default(),
                attributes: item.attrs.clone(),
            },
        );

        for attr in syn::punctuated::Punctuated::<syn::Ident, syn::Token![,]>::parse_terminated
            .parse2(attrs)
            .unwrap()
        {

                output.0.insert(attr.to_string(), Default::default());
        }

        Ok(output)
    }

    pub fn get_mut(&mut self, name: &str) -> Option<&mut OutputStruct> {
        self.0.get_mut(name)
    }
}

impl OutputStruct {
    pub fn missing_fields_from(&self, other: &Self) -> Vec<OutputField> {
        self.fields.iter().fold(vec![], |mut acc, field| {
            if !other.fields.contains(field) {
                acc.push(field.clone())
            }
            acc
        })
    }

    pub fn same_fields_as(&self, other: &Self) -> Vec<OutputField> {
        self.fields.iter().fold(vec![], |mut acc, field| {
            if other.fields.contains(field) {
                acc.push(field.clone())
            }
            acc
        })
    }
}

impl OutputField {
    pub fn name(&self) -> syn::Ident {
        self.definition
            .ident
            .clone()
            .unwrap_or_else(|| panic!("Can't get field name. This should never happen."))
    }

    pub fn trait_name(&self) -> syn::Ident {
        syn::Ident::new(
            &format!("Has{}", &self.name().to_string().to_pascal_case()),
            proc_macro2::Span::call_site(),
        )
    }

    pub fn neg_trait_name(&self) -> syn::Ident {
        syn::Ident::new(
            &format!("HasNo{}", &self.name().to_string().to_pascal_case()),
            proc_macro2::Span::call_site(),
        )
    }
}

impl PartialEq for OutputField {
    fn eq(&self, other: &Self) -> bool {
        self.name() == other.name()
    }
}

impl From<syn::Field> for OutputField {
    fn from(field: syn::Field) -> Self {
        Self::new(field, false)
    }
}

impl From<OutputField> for syn::Field {
    fn from(field_config: OutputField) -> Self {
        field_config.definition
    }
}
