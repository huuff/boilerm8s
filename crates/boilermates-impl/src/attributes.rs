use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse_quote, Attribute, Lit, Meta, MetaList};

nestify::nest! {
  pub enum BoilermatesStructAttribute {
    AttrFor(pub struct BoilermatesAttrFor {
      pub target_struct: String,
      pub attribute: Attribute,
    })
  }
}

impl BoilermatesStructAttribute {
    pub fn extract(attributes: &mut Vec<Attribute>) -> Result<Vec<Self>, anyhow::Error> {
        use itertools::Itertools as _;

        let (boilermates_attrs, non_boilermates_attrs) = std::mem::take(attributes)
            .into_iter()
            .partition(is_boilermates);

        *attributes = non_boilermates_attrs;

        boilermates_attrs
            .into_iter()
            .map(|attr| match &attr.meta {
                Meta::List(list) => {
                    let meta: Meta = list.parse_args()?;
                    match meta {
                        Meta::List(attr) if attr.path.is_ident("attr_for") => {
                            Ok(BoilermatesStructAttribute::AttrFor(
                                BoilermatesAttrFor::try_from(attr)?,
                            ))
                        }
                        _ => Err(anyhow::anyhow!(
                            "unknown boilermates attribute: `{}`",
                            quote::quote!(#meta)
                        )),
                    }
                }
                _ => Err(anyhow::anyhow!(
                    "unknown boilermates attribute: `{}`",
                    quote::quote!(#attr)
                )),
            })
            .try_collect()
    }
}

fn is_boilermates(attr: &Attribute) -> bool {
    attr.path().is_ident("boilermates")
}

impl TryFrom<MetaList> for BoilermatesAttrFor {
    type Error = anyhow::Error;

    fn try_from(list_attr: MetaList) -> Result<Self, Self::Error> {
        use syn::parse::Parser;
        use syn::punctuated::Punctuated;

        let nested =
            Punctuated::<Lit, syn::Token![,]>::parse_terminated.parse2(list_attr.tokens)?;

        if nested.len() != 2 {
            anyhow::bail!("`#[boilermates(attr_for(...))]` must have two string literal arguments");
        }

        let mut list_iter = nested.into_iter();

        match (
            list_iter.next().expect("we just checked length is 2"),
            list_iter.next().expect("we just checked length is 2"),
        ) {
            (Lit::Str(struct_name), Lit::Str(attr_list)) => Ok(Self {
                target_struct: struct_name.value(),
                attribute: {
                    let token_stream: TokenStream = attr_list
                        .value()
                        .parse::<TokenStream>()
                        .map_err(|_| anyhow::anyhow!("can't parse attr"))?;
                    let q = quote! { #token_stream };
                    parse_quote! { #q }
                },
            }),
            _ => anyhow::bail!(
                "`#[boilermates(attr_for(...))]` must have two string literal arguments"
            ),
        }
    }
}

nestify::nest! {
  pub enum BoilermatesFieldAttribute {
    OnlyIn(pub struct BoilermatesOnlyIn(pub Vec<String>)),
    NotIn(pub struct BoilermatesNotIn(pub Vec<String>)),
    Default,
    OnlyInSelf,
  }
}

impl BoilermatesFieldAttribute {
    pub fn extract(
        attributes: &mut Vec<Attribute>,
    ) -> Result<Vec<BoilermatesFieldAttribute>, anyhow::Error> {
        use itertools::Itertools as _;

        let (boilermates_attrs, non_boilermates_attrs) = std::mem::take(attributes)
            .into_iter()
            .partition(is_boilermates);

        *attributes = non_boilermates_attrs;

        boilermates_attrs
            .into_iter()
            .map(|attr| match &attr.meta {
                Meta::List(list) => {
                    let meta: Meta = list.parse_args()?;
                    match meta {
                        Meta::List(list) if list.path.is_ident("only_in") => {
                            Ok(BoilermatesFieldAttribute::OnlyIn(BoilermatesOnlyIn(
                                extract_nested_list(&list)?,
                            )))
                        }
                        Meta::List(list) if list.path.is_ident("not_in") => {
                            Ok(BoilermatesFieldAttribute::NotIn(BoilermatesNotIn(
                                extract_nested_list(&list)?,
                            )))
                        }
                        Meta::Path(path) if path.is_ident("default") => {
                            Ok(BoilermatesFieldAttribute::Default)
                        }
                        Meta::Path(path) if path.is_ident("only_in_self") => {
                            Ok(BoilermatesFieldAttribute::OnlyInSelf)
                        }
                        _ => anyhow::bail!(
                            "unknown boilermates attribute: `{}`",
                            quote::quote!(#meta)
                        ),
                    }
                }
                _ => Err(anyhow::anyhow!(
                    "unknown boilermates attribute: `{}`",
                    quote::quote!(#attr)
                )),
            })
            .try_collect()
    }
}

fn extract_nested_list(meta_list: &MetaList) -> anyhow::Result<Vec<String>> {
    use syn::parse::Parser;
    use syn::punctuated::Punctuated;

    let nested =
        Punctuated::<Lit, syn::Token![,]>::parse_terminated.parse2(meta_list.tokens.clone())?;
    nested
        .into_iter()
        .map(|n| match n {
            Lit::Str(lit) => Ok(lit.value()),
            _ => anyhow::bail!("Expected a string literal"),
        })
        .collect()
}
