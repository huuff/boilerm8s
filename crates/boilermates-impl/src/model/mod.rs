pub mod output;
pub mod fields;