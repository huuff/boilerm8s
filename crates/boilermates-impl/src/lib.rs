mod attributes;
mod model;
mod util;

use attributes::{BoilermatesFieldAttribute, BoilermatesStructAttribute};
use heck::ToSnakeCase;
use itertools::Itertools;
use model::{OutputField, OutputStructs};
use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{Data, DataStruct, DeriveInput, Field, Fields, FieldsNamed};

pub fn boilermates(attrs: TokenStream2, item: TokenStream2) -> TokenStream2 {
    let mut item: DeriveInput = syn::parse2(item).unwrap();

    // XXX: must do this first to remove all boilermates attributes before initializing the output
    let stuct_attrs = BoilermatesStructAttribute::extract(&mut item.attrs).unwrap();
    let mut structs = OutputStructs::initialize(&item, attrs).unwrap();

    let Data::Struct(data_struct) = item.data.clone() else {
        panic!("Expected a struct");
    };

    let Fields::Named(mut fields) = data_struct.fields.clone() else {
        panic!("Expected a struct with named fields");
    };

    for struct_attr in stuct_attrs {
        match struct_attr {
            BoilermatesStructAttribute::AttrFor(attr_for) => {
                structs
                    .get_mut(&attr_for.target_struct)
                    .unwrap()
                    .attributes
                    .push(attr_for.attribute);
            }
        }
    }

    let mut traits = quote! {};

    fields.named.iter_mut().for_each(|field| {
        let mut add_to = structs.names().map(<_>::to_owned).collect_vec();
        let mut default = false;

        for boilermates_attr in BoilermatesFieldAttribute::extract(&mut field.attrs).unwrap() {
            match boilermates_attr {
                BoilermatesFieldAttribute::OnlyIn(only_in) => add_to = only_in.0,
                BoilermatesFieldAttribute::NotIn(not_in) => {
                    add_to.retain(|strukt| !not_in.0.contains(&strukt))
                }
                BoilermatesFieldAttribute::Default => default = true,
                BoilermatesFieldAttribute::OnlyInSelf => add_to = vec![item.ident.to_string()],
            }
        }

        let field = OutputField::new(field.clone(), default);
        let trait_name = field.trait_name();
        let neg_trait_name = field.neg_trait_name();
        let field_name = field.name();
        let setter_fn = Ident::new(&format!("set_{}", field_name), Span::call_site());
        let field_ty = &field.definition.ty;
        traits = quote! {
            #traits
            trait #trait_name {
                fn #field_name(&self) -> &#field_ty;
                fn #setter_fn(&mut self, value: #field_ty);
            }

            trait #neg_trait_name {}
        };

        for (struct_name, strukt) in &mut structs {
            // MAYBE we don't need to store names and strings and can store idents directly?
            let struct_ident = Ident::new(struct_name, Span::call_site());

            if add_to.contains(struct_name) {
                strukt.fields.push(field.clone());

                traits = quote! {
                    #traits
                    impl #trait_name for #struct_ident {
                        fn #field_name(&self) -> &#field_ty {
                            &self.#field_name
                        }

                        fn #setter_fn(&mut self, value: #field_ty) {
                            self.#field_name = value;
                        }
                    }
                };
            } else {
                traits = quote! {
                    #traits
                    impl #neg_trait_name for #struct_ident {}
                };
            }
        }
    });

    let mut output = quote! {};
    for (name, strukt) in &structs {
        let out_struct = DeriveInput {
            attrs: strukt.attributes.clone(),
            data: Data::Struct(DataStruct {
                fields: Fields::Named(FieldsNamed {
                    named: strukt
                        .fields
                        .iter()
                        .cloned()
                        .map(Into::<Field>::into)
                        .collect(),
                    ..fields
                }),
                ..data_struct
            }),
            ident: Ident::new(name, Span::call_site()),
            ..item.clone()
        };
        output = quote! {
            #output
            #out_struct
        };

        for (other_name, other) in &structs {
            if name == other_name {
                continue;
            }
            let name = Ident::new(name, Span::call_site());
            let other_name = Ident::new(other_name, Span::call_site());
            let missing_fields = strukt.missing_fields_from(other);
            let missing_fields_without_defaults = missing_fields
                .iter()
                .filter(|f| !f.default)
                .collect::<Vec<_>>();

            let default_field_setters =
                missing_fields
                    .iter()
                    .filter(|f| f.default)
                    .fold(quote! {}, |acc, field| {
                        let field_name = field.name();
                        quote! {
                            #acc
                            #field_name: Default::default(),
                        }
                    });

            if missing_fields_without_defaults.is_empty() {
                let common_field_setters =
                    strukt
                        .same_fields_as(other)
                        .iter()
                        .fold(quote! {}, |acc, field| {
                            let field_name = &field.name();
                            quote! {
                                #acc
                                #field_name: other.#field_name,
                            }
                        });

                output = quote! {
                    #output
                    impl From<#other_name> for #name {
                        fn from(other: #other_name) -> Self {
                            Self {
                                #common_field_setters
                                #default_field_setters
                            }
                        }
                    }
                };
            }
            if !missing_fields.is_empty() {
                let common_field_setters =
                    strukt
                        .same_fields_as(other)
                        .iter()
                        .fold(quote! {}, |acc, field| {
                            let field_name = field.name();
                            quote! {
                                #acc
                                #field_name: self.#field_name,
                            }
                        });

                let into_args = missing_fields.iter().fold(quote! {}, |acc, field| {
                    let field_name = field.name();
                    let field_ty = &field.definition.ty;
                    quote! {
                        #acc
                        #field_name: #field_ty,
                    }
                });

                let into_defaults_args =
                    missing_fields_without_defaults
                        .iter()
                        .fold(quote! {}, |acc, field| {
                            let field_name = field.name();
                            let field_ty = &field.definition.ty;
                            quote! {
                                #acc
                                #field_name: #field_ty,
                            }
                        });

                let into_missing_setters = missing_fields.iter().fold(quote! {}, |acc, field| {
                    let field_name = field.name();
                    quote! { #acc #field_name, }
                });

                let into_defaults_missing_setters =
                    missing_fields_without_defaults
                        .iter()
                        .fold(quote! {}, |acc, field| {
                            let field_name = field.name();
                            quote! { #acc #field_name, }
                        });

                let into_defaults_fn_name = Ident::new(
                    &format!("into{}_defaults", name).to_snake_case(),
                    Span::call_site(),
                );

                let into_fn_name =
                    Ident::new(&format!("into{}", name).to_snake_case(), Span::call_site());

                output = quote! {
                    #output
                    impl #other_name {
                        pub fn #into_fn_name(self, #into_args) -> #name {
                            #name {
                                #common_field_setters
                                #into_missing_setters
                            }
                        }

                        pub fn #into_defaults_fn_name(self, #into_defaults_args) -> #name {
                            #name {
                                #common_field_setters
                                #default_field_setters
                                #into_defaults_missing_setters
                            }
                        }
                    }
                };
            }
        }
    }

    output = quote! {
        #output
        #traits
    };

    output.into()
}

#[cfg(test)]
mod test {
    use crate::util::pretty_print;

    use super::*;
    use quote::quote;

    #[test]
    fn snapshot_test() {
        let output = boilermates(
            quote! { StructWithX, StructWithoutY },
            quote! {
              pub struct MainStruct {
                pub field: String,
                #[boilermates(only_in(StructWithX))]
                pub x: u32,
                #[boilermates(not_in(StructWithoutY))]
                pub y: i32,
              }
            },
        );

        insta::assert_snapshot!(pretty_print(output));
    }
}
