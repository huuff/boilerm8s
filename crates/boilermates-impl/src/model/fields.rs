use crate::attributes::BoilermatesFieldAttribute;

#[derive(Clone)]
pub struct BoilermatesField {
    pub definition: syn::Field,
    pub default: bool,
    pub in_structs: Vec<String>,
}

impl BoilermatesField {
    pub fn parse_all(
        syn_fields: syn::punctuated::Punctuated<syn::Field, syn::token::Comma>,
        all_boilermates_structs: Vec<String>,
        main_boilermates_struct: String,
    ) -> Vec<Self> {
        syn_fields
            .into_iter()
            .map(|mut syn_field| {
                let mut default = false;
                let mut in_structs = all_boilermates_structs.clone();

                for boilermates_attr in
                    BoilermatesFieldAttribute::extract(&mut syn_field.attrs).unwrap()
                {
                    match boilermates_attr {
                        BoilermatesFieldAttribute::OnlyIn(only_in) => in_structs = only_in.0,
                        BoilermatesFieldAttribute::NotIn(not_in) => {
                            in_structs.retain(|strukt| !not_in.0.contains(&strukt))
                        }
                        BoilermatesFieldAttribute::Default => default = true,
                        BoilermatesFieldAttribute::OnlyInSelf => {
                            in_structs = vec![main_boilermates_struct.clone()]
                        }
                    }
                }

                Self {
                    definition: syn_field,
                    default,
                    in_structs,
                }
            })
            .collect()
    }

    pub fn name(&self) -> syn::Ident {
        self.definition
            .ident
            .clone()
            .unwrap_or_else(|| panic!("Can't get field name. This should never happen."))
    }
}
