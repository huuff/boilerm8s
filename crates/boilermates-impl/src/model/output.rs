use derive_more::{Constructor, IntoIterator};
use indexmap::IndexMap;

use super::fields::BoilermatesField;

nestify::nest! {
  // XXX: Use an indexmap so it's insertion order and deterministic
  #[derive(Default, IntoIterator)]
  pub struct OutputStructs(
    #[into_iterator(owned, ref, ref_mut)]
    #>[derive(Default)]
    pub IndexMap<String, pub struct OutputStruct {
      pub attributes: Vec<syn::Attribute>,
      #>[derive(Clone, Constructor)]
      pub fields: Vec<pub struct OutputField {
        pub definition: syn::Field,
        pub default: bool,
      }>,
  }>)
}

impl OutputStructs {
    /// Returns the names of all the output structs
    pub fn names(&self) -> impl Iterator<Item = &str> {
        self.0.keys().map(|it| it.as_str())
    }

    /// Initializes the output with the main struct and all the mates
    pub fn initialize(
        item: &syn::DeriveInput,
        attrs: proc_macro2::TokenStream,
    ) -> Result<Self, anyhow::Error> {
        use syn::parse::Parser as _;

        let mut output = Self::default();

        output.0.insert(
            item.ident.to_string(),
            OutputStruct {
                fields: Default::default(),
                attributes: item.attrs.clone(),
            },
        );

        for attr in syn::punctuated::Punctuated::<syn::Ident, syn::Token![,]>::parse_terminated
            .parse2(attrs)
            .unwrap()
        {
            output.0.insert(attr.to_string(), Default::default());
        }

        Ok(output)
    }

    pub fn get_mut(&mut self, name: &str) -> Option<&mut OutputStruct> {
        self.0.get_mut(name)
    }
}

impl OutputStruct {
    pub fn missing_fields_from(&self, other: &Self) -> Vec<OutputField> {
        self.fields.iter().fold(vec![], |mut acc, field| {
            if !other.fields.contains(field) {
                acc.push(field.clone())
            }
            acc
        })
    }

    pub fn same_fields_as(&self, other: &Self) -> Vec<OutputField> {
        self.fields.iter().fold(vec![], |mut acc, field| {
            if other.fields.contains(field) {
                acc.push(field.clone())
            }
            acc
        })
    }
}

impl From<BoilermatesField> for OutputField {
    fn from(value: BoilermatesField) -> Self {
        Self::new(value.definition, value.default)
    }
}

impl OutputField {
    // TODO: Pretty much copy-paste from `BoilermatesField`, DRY?
    pub fn name(&self) -> syn::Ident {
        self.definition
            .ident
            .clone()
            .unwrap_or_else(|| panic!("Can't get field name. This should never happen."))
    }
}

impl PartialEq for OutputField {
    fn eq(&self, other: &Self) -> bool {
        self.name() == other.name()
    }
}

impl From<syn::Field> for OutputField {
    fn from(field: syn::Field) -> Self {
        Self::new(field, false)
    }
}

impl From<OutputField> for syn::Field {
    fn from(field_config: OutputField) -> Self {
        field_config.definition
    }
}
