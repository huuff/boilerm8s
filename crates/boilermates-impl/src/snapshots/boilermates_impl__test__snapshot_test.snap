---
source: crates/boilermates-impl/src/lib.rs
expression: pretty_print(output)
---
pub struct MainStruct {
    pub field: String,
    pub y: i32,
}
impl From<StructWithX> for MainStruct {
    fn from(other: StructWithX) -> Self {
        Self {
            field: other.field,
            y: other.y,
        }
    }
}
impl StructWithoutY {
    pub fn into_main_struct(self, y: i32) -> MainStruct {
        MainStruct { field: self.field, y }
    }
    pub fn into_main_struct_defaults(self, y: i32) -> MainStruct {
        MainStruct { field: self.field, y }
    }
}
pub struct StructWithX {
    pub field: String,
    pub x: u32,
    pub y: i32,
}
impl MainStruct {
    pub fn into_struct_with_x(self, x: u32) -> StructWithX {
        StructWithX {
            field: self.field,
            y: self.y,
            x,
        }
    }
    pub fn into_struct_with_x_defaults(self, x: u32) -> StructWithX {
        StructWithX {
            field: self.field,
            y: self.y,
            x,
        }
    }
}
impl StructWithoutY {
    pub fn into_struct_with_x(self, x: u32, y: i32) -> StructWithX {
        StructWithX {
            field: self.field,
            x,
            y,
        }
    }
    pub fn into_struct_with_x_defaults(self, x: u32, y: i32) -> StructWithX {
        StructWithX {
            field: self.field,
            x,
            y,
        }
    }
}
pub struct StructWithoutY {
    pub field: String,
}
impl From<MainStruct> for StructWithoutY {
    fn from(other: MainStruct) -> Self {
        Self { field: other.field }
    }
}
impl From<StructWithX> for StructWithoutY {
    fn from(other: StructWithX) -> Self {
        Self { field: other.field }
    }
}
trait HasField {
    fn field(&self) -> &String;
    fn set_field(&mut self, value: String);
}
trait HasNoField {}
impl HasField for MainStruct {
    fn field(&self) -> &String {
        &self.field
    }
    fn set_field(&mut self, value: String) {
        self.field = value;
    }
}
impl HasField for StructWithX {
    fn field(&self) -> &String {
        &self.field
    }
    fn set_field(&mut self, value: String) {
        self.field = value;
    }
}
impl HasField for StructWithoutY {
    fn field(&self) -> &String {
        &self.field
    }
    fn set_field(&mut self, value: String) {
        self.field = value;
    }
}
trait HasX {
    fn x(&self) -> &u32;
    fn set_x(&mut self, value: u32);
}
trait HasNoX {}
impl HasNoX for MainStruct {}
impl HasX for StructWithX {
    fn x(&self) -> &u32 {
        &self.x
    }
    fn set_x(&mut self, value: u32) {
        self.x = value;
    }
}
impl HasNoX for StructWithoutY {}
trait HasY {
    fn y(&self) -> &i32;
    fn set_y(&mut self, value: i32);
}
trait HasNoY {}
impl HasY for MainStruct {
    fn y(&self) -> &i32 {
        &self.y
    }
    fn set_y(&mut self, value: i32) {
        self.y = value;
    }
}
impl HasY for StructWithX {
    fn y(&self) -> &i32 {
        &self.y
    }
    fn set_y(&mut self, value: i32) {
        self.y = value;
    }
}
impl HasNoY for StructWithoutY {}
